In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from datetime import datetime
import datetime as datetime
import warnings
warnings.filterwarnings('ignore')
import pandas 

In [2]:
contents=pd.read_csv('../orga2_21/contents.csv',low_memory=False)

In [3]:
contents.head()

title  id  namespace  revision_id  \
0  Wikipedia:Artículos solicitados   5          4    132533307   
1                          Andorra   7          0    138035057   
2                        Argentina  10          0    137966826   
3             Geografía de Andorra  15          0    137986788   
4            Demografía de Andorra  17          0    130869554   

   parent_revision    revision_timestamp revisor_username  revisor_id  \
0      132516631.0  2021-01-19T18:40:34Z        Metrónomo   1708233.0   
1      138027475.0  2021-08-31T18:27:01Z              NaN         NaN   
2      137825165.0  2021-08-28T17:17:54Z        Traitrich   4642283.0   
3      133666383.0  2021-08-29T17:16:00Z            Lojwe   5338508.0   
4      129413188.0  2020-11-12T19:56:15Z           Jialxv   3615366.0   

                               revisor_ip  \
0                                     NaN   
1  2800:CD0:FC1B:A900:C5A4:9E81:7175:2ED2   
2                                     NaN   
3                                     NaN   
4                                     NaN   

                                     revisor_comment  
0                                     Así esta mejor  
1                                   Añadir contenido  
2  Reemplazos con [[Usuario:Benjavalero/Replacer|...  
3  Reemplazos con [[Usuario:Benjavalero/Replacer|...  
4  Correcciones ortográficas con [[Usuario:Benjav...

In [4]:
contents.dtypes

title                  object
id                      int64
namespace               int64
revision_id             int64
parent_revision       float64
revision_timestamp     object
revisor_username       object
revisor_id            float64
revisor_ip             object
revisor_comment        object
dtype: object

In [5]:
#cambiamos el formato de timestamp
contents.loc[:,'revision_timestamp']=pd.to_datetime(contents['revision_timestamp'],errors='coerce')

In [6]:
contents.dtypes

title                              object
id                                  int64
namespace                           int64
revision_id                         int64
parent_revision                   float64
revision_timestamp    datetime64[ns, UTC]
revisor_username                   object
revisor_id                        float64
revisor_ip                         object
revisor_comment                    object
dtype: object

In [7]:
contents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4132164 entries, 0 to 4132163
Data columns (total 10 columns):
 #   Column              Dtype              
---  ------              -----              
 0   title               object             
 1   id                  int64              
 2   namespace           int64              
 3   revision_id         int64              
 4   parent_revision     float64            
 5   revision_timestamp  datetime64[ns, UTC]
 6   revisor_username    object             
 7   revisor_id          float64            
 8   revisor_ip          object             
 9   revisor_comment     object             
dtypes: datetime64[ns, UTC](1), float64(2), int64(3), object(4)
memory usage: 315.3+ MB


In [8]:
contents.isnull().sum()

title                       4
id                          0
namespace                   0
revision_id                 0
parent_revision       1483218
revision_timestamp          0
revisor_username       193287
revisor_id             193287
revisor_ip            3938878
revisor_comment        469224
dtype: int64

__* Como tenemos 4 nulos en title de 4132164 registros entonces se pueden eliminar directamente__

In [9]:
contents.dropna(subset=['title'], inplace=True)

In [10]:
contents.isnull().sum()

title                       0
id                          0
namespace                   0
revision_id                 0
parent_revision       1483217
revision_timestamp          0
revisor_username       193287
revisor_id             193287
revisor_ip            3938874
revisor_comment        469224
dtype: int64

In [11]:
print(contents.shape)

(4132160, 10)


*** 
***
#  Ejercicio 7. 
#  La antigüedad promedio de la última edición de los artículos cuyo título contenga tu apellido (si no hay, tu nombre y si tampoco hay usa Cafferata) 
***
***

In [58]:
antiguedad_apenom=contents[['title','revision_timestamp']]

In [59]:
antiguedad_apenom.head()

title        revision_timestamp
0  Wikipedia:Artículos solicitados 2021-01-19 18:40:34+00:00
1                          Andorra 2021-08-31 18:27:01+00:00
2                        Argentina 2021-08-28 17:17:54+00:00
3             Geografía de Andorra 2021-08-29 17:16:00+00:00
4            Demografía de Andorra 2020-11-12 19:56:15+00:00

In [60]:
antiguedad_apenom.dtypes

title                              object
revision_timestamp    datetime64[ns, UTC]
dtype: object

In [61]:
antiguedad_apenom.loc[:,'revision_timestamp']=pd.to_datetime(antiguedad_apenom['revision_timestamp'].dt.strftime("%d/%m/%Y"),errors='coerce')
antiguedad_apenom.head()

title revision_timestamp
0  Wikipedia:Artículos solicitados         2021-01-19
1                          Andorra         2021-08-31
2                        Argentina         2021-08-28
3             Geografía de Andorra         2021-08-29
4            Demografía de Andorra         2020-12-11

In [62]:
filtroXapellido=antiguedad_apenom
filtroXapellido=filtroXapellido[filtroXapellido.title.str.contains('Porras',case=False)]
filtroXapellido.reset_index(drop=True,inplace=True)
filtroXapellido.head(20)

title revision_timestamp
0           Miguel Mora Porras         2021-08-27
1      Juan Rafael Mora Porras         2021-08-27
2        Línea Anderson-Porras         2019-10-11
3   Convención Anderson-Porras         2019-08-30
4             Martín de Porras         2006-06-17
5      Raúl Porras Barrenechea         2021-02-08
6             Belisario Porras         2021-07-13
7        José Francisco Porras         2021-07-19
8                  La Porrassa         2021-12-07
9     José Joaquín Mora Porras         2021-08-27
10           Luis Gallo Porras         2021-08-21
11     Lizardo Alzamora Porras         2021-08-31
12          Porras barrenechea         2007-08-07
13     Gustavo Porras Castejón         2020-05-16
14  José Gregorio Bello Porras         2021-04-02
15          Jorge Conte Porras         2021-06-25
16   Belisario porras barahona         2010-11-03
17  Convencion Anderson-Porras         2008-03-11
18     Gustavo Porras Castejon         2008-11-27
19       Jose Francisco Porras         2008-05-12

In [63]:
antiguedad_apenom.head()

title revision_timestamp
0  Wikipedia:Artículos solicitados         2021-01-19
1                          Andorra         2021-08-31
2                        Argentina         2021-08-28
3             Geografía de Andorra         2021-08-29
4            Demografía de Andorra         2020-12-11

In [64]:
print(filtroXapellido.shape)

(138, 2)


In [65]:
print(antiguedad_apenom.shape)

(4132160, 2)


In [66]:
#Para calcular la diferencia de dias entre revision_timestamp y la fecha de hoy
def date_diff(date1, date2):
    return ((date1 -  date2).dt.days)
   

In [67]:
filtroXapellido['difDays']=date_diff(datetime.datetime.today(),filtroXapellido.loc[:,'revision_timestamp'])

In [68]:
filtroXapellido.head()

title revision_timestamp  difDays
0          Miguel Mora Porras         2021-08-27       46
1     Juan Rafael Mora Porras         2021-08-27       46
2       Línea Anderson-Porras         2019-10-11      732
3  Convención Anderson-Porras         2019-08-30      774
4            Martín de Porras         2006-06-17     5596

In [69]:
promedioDifDays=filtroXapellido['difDays'].mean()
promedioDifDays

1750.6304347826087

In [70]:
promDifAnios=promedioDifDays/365.2425
promDifAnios=int(promDifAnios)
promDifAnios

4

In [71]:
a=(filtroXapellido['difDays']/365.2425).mean()

In [72]:
int(a)

4

# Rpta: La antigüedad promedio de la ultima revision de los articulos es 4 años

*** 
***
#  Ejercicio 8. 
#  La mediana de la antigüedad para las últimas ediciones vigentes agrupado por el primer carácter del título 
***
***

* __En la columna "firstChar" tendremos al primer caracter de los titulos__

In [73]:
mediana_caracter=antiguedad_apenom
mediana_caracter['firstChar']=mediana_caracter['title'].str.get(0)
mediana_caracter.head()

title revision_timestamp firstChar
0  Wikipedia:Artículos solicitados         2021-01-19         W
1                          Andorra         2021-08-31         A
2                        Argentina         2021-08-28         A
3             Geografía de Andorra         2021-08-29         G
4            Demografía de Andorra         2020-12-11         D

In [74]:
print(mediana_caracter.shape)

(4132160, 3)


In [75]:
mediana_caracter.firstChar.value_counts()

C    818489
A    380186
P    297262
M    282296
S    252335
      ...  
ო         1
Ơ         1
𐤊         1
ざ         1
虍         1
Name: firstChar, Length: 1846, dtype: int64

* __Ahora creamos una columna donde se obtenga la diferencia de dias que existe entre la ultima revisión y el dia de hoy así tendremos un promedio de dias o de años.__

In [97]:
mediana_caracter['difDays']=date_diff(datetime.datetime.today(),mediana_caracter.loc[:,'revision_timestamp'])
mediana_caracter.head()

title revision_timestamp firstChar  difDays
0  Wikipedia:Artículos solicitados         2021-01-19         W      266
1                          Andorra         2021-08-31         A       42
2                        Argentina         2021-08-28         A       45
3             Geografía de Andorra         2021-08-29         G       44
4            Demografía de Andorra         2020-12-11         D      305

In [98]:
final=mediana_caracter[['firstChar','difDays']]

In [99]:
final.head()

firstChar  difDays
0         W      266
1         A       42
2         A       45
3         G       44
4         D      305

In [100]:
final=final.groupby(['firstChar']).agg({'difDays':['mean','count','sum']})

In [101]:
final.reset_index(inplace=True)
final.head()

firstChar     difDays            
                   mean count   sum
0         !  210.700000    10  2107
1         "  439.090909    22  9660
2         $  300.166667     6  1801
3         %   86.500000     2   173
4         &  458.500000     6  2751

In [102]:
final.columns

MultiIndex([('firstChar',      ''),
            (  'difDays',  'mean'),
            (  'difDays', 'count'),
            (  'difDays',   'sum')],
           )

In [103]:
final.columns=['firstChar','mean_difDays','count_difDays','sum_difDays']
final.head()

firstChar  mean_difDays  count_difDays  sum_difDays
0         !    210.700000             10         2107
1         "    439.090909             22         9660
2         $    300.166667              6         1801
3         %     86.500000              2          173
4         &    458.500000              6         2751

* __Como podemos observar tenemos el promedio de diferencia de días para cada caracter asi como su cantidad y com la suma. Ademas mostraremos la antiguedad en años__

In [104]:
final['mean_anios']=(final['mean_difDays']/365.2425).astype(np.int16)
final.head(20)

firstChar  mean_difDays  count_difDays  sum_difDays  mean_anios
0          !    210.700000             10         2107           0
1          "    439.090909             22         9660           1
2          $    300.166667              6         1801           0
3          %     86.500000              2          173           0
4          &    458.500000              6         2751           1
5          '    681.698795             83        56581           1
6          (    793.177053          12516      9927404           2
7          )   3679.500000              2         7359          10
8          *   2908.521739             23        66896           7
9          +   1974.468085             47        92800           5
10         ,   3240.000000              1         3240           8
11         -   1911.906250             64       122362           5
12         .   1262.809593            688       868813           3
13         /   1744.692308             26        45362           4
14         0   3035.989437            284       862221           8
15         1   1803.775077           8394     15140888           4
16         2   2065.498779           5732     11839439           5
17         3   2294.254624           3460      7938121           6
18         4   2375.186126           2912      6916542           6
19         5   2805.193562           2299      6449140           7

* __En lo siguiente se muestra los caracteres que tuvieron mayor aparición como el primer carácter de title.__

In [111]:
final.nlargest(50,'count_difDays')

firstChar  mean_difDays  count_difDays  sum_difDays  mean_anios
30          C   1840.416386         818489   1506360567           5
28          A   1730.406122         380186    657876182           4
43          P   2012.604003         297262    598270691           5
40          M   1839.186584         282296    519195016           5
46          S   2035.393513         252335    513601022           5
32          E   1980.444109         208699    413316705           5
39          L   1992.717954         189526    377671863           5
47          T   1838.803652         165736    304755962           5
45          R   1912.911297         160119    306293444           5
29          B   2050.806301         159376    326849305           5
37          J   1990.571645         145377    289383334           5
31          D   1894.833616         135794    257307036           5
33          F   1996.201277         120128    239799667           5
34          G   1948.163611         119845    233477668           5
35          H   1875.775549         101314    190042324           5
36          I   1972.747502          94369    186166209           5
50          W   2023.603598          78375    158599932           5
41          N   1859.267627          75284    139973104           5
49          V   2022.125999          74429    150504816           5
42          O   1937.429223          62266    120635968           5
38          K   1566.103282          51577     80774909           4
48          U   2050.489457          37133     76140825           5
52          Y   1781.442007          18459     32883638           4
53          Z   1865.280834          16362     30519725           5
6           (    793.177053          12516      9927404           2
44          Q   2106.481616          10172     21427131           5
82          Á   1924.087594           9076     17463019           5
51          X   1964.029671           8729     17144015           5
15          1   1803.775077           8394     15140888           4
16          2   2065.498779           5732     11839439           5
17          3   2294.254624           3460      7938121           6
90          É   1930.357099           3254      6281382           5
18          4   2375.186126           2912      6916542           6
100         Ó   1900.640620           2905      5521361           5
19          5   2805.193562           2299      6449140           7
94          Í   1989.315982           1364      2713427           5
20          6   2091.417957           1292      2702112           5
80          ¿   2619.166937           1234      3232052           7
58          ¡   2253.877223           1181      2661829           6
21          7   1912.619130           1150      2199512           5
22          8   1755.786683            811      1423943           4
23          9   1753.297980            792      1388612           4
12          .   1262.809593            688       868813           3
158         Š   1219.667826            575       701309           3
106         Ú   1471.717949            507       746161           4
119         Č   1297.502994            501       650049           3
172         Ž   1098.400000            420       461328           3
110         Þ   2081.819149            376       782764           5
155         Ś   2706.685629            334       904033           7
103         Ö   1719.283784            296       508908           4